### ISSUE TITLE
Failure specifying only target name containing unusual characters

### ISSUE THREAD

**Issue by trishayy** [realm/jazzy#422]
tldr; We think `xcodebuild` changes hyphens to underscores in the module name when it reports its output.

We have an iOS project written in Swift which we want to document with Jazzy. Here is how we run Jazzy: 

```
jazzy -o /Users/milbuild/jenkins-build/workspace/raDocs_CI/js/gulp-jazzy/../../gen/auto/apps/venue_1_0_0/jazzy/ios --source-directory . --min-acl private -x '-target,Venue-companion'
```

When we run this command we get this error:

```
Running xcodebuild
Could not parse compiler arguments from `xcodebuild` output.
Please confirm that `xcodebuild` is building a Swift module.
Saved `xcodebuild` log file: /var/folders/hr/94_zbrf53jz4jtmyqr6_0p4r0000gn/T/xcodebuild-137485DB-6E27-450D-A87C-38C165681293.log
Failed to generate documentation
/usr/local/lib/ruby/gems/2.2.0/gems/jazzy-0.3.2/lib/jazzy/executable.rb:35:in `execute_command': /usr/local/lib/ruby/gems/2.2.0/gems/jazzy-0.3.2/lib/jazzy/SourceKitten/bin/sourcekitten ["doc", "-target", "Venue-companion"] (RuntimeError)

Running xcodebuild

Could not parse compiler arguments from `xcodebuild` output.

Please confirm that `xcodebuild` is building a Swift module.

Saved `xcodebuild` log file: /var/folders/hr/94_zbrf53jz4jtmyqr6_0p4r0000gn/T/xcodebuild-137485DB-6E27-450D-A87C-38C165681293.log

Failed to generate documentation
    from /usr/local/lib/ruby/gems/2.2.0/gems/jazzy-0.3.2/lib/jazzy/sourcekitten.rb:128:in `run_sourcekitten'
    from /usr/local/lib/ruby/gems/2.2.0/gems/jazzy-0.3.2/lib/jazzy/doc_builder.rb:56:in `block in build'
    from /usr/local/lib/ruby/gems/2.2.0/gems/jazzy-0.3.2/lib/jazzy/doc_builder.rb:54:in `chdir'
    from /usr/local/lib/ruby/gems/2.2.0/gems/jazzy-0.3.2/lib/jazzy/doc_builder.rb:54:in `build'
    from /usr/local/lib/ruby/gems/2.2.0/gems/jazzy-0.3.2/bin/jazzy:15:in `<top (required)>'
    from /usr/local/bin/jazzy:23:in `load'
    from /usr/local/bin/jazzy:23:in `<main>'
```

We didn't think we were doing anything wrong because we run this exact same command successfully on different projects with different target names that do not have a hyphen. For example:

```
jazzy -o /Users/milbuild/jenkins-build/workspace/raDocs_CI/js/gulp-jazzy/../../gen/auto/apps/venue_1_0_0/jazzy/ios --source-directory . --min-acl private -x '-target,Travel'
```

We resorted to looking at the SourceKitten source code. We searched the project for `Could not parse compiler arguments from xcodebuild output.` which brought us to [this file](https://github.com/jpsim/SourceKitten/blob/f9b95146955da3376b015ac21df6eb9cee7b8ca7/Source/SourceKittenFramework/Module.swift). 

First, we found the `xcodebuild` command that SourceKitten is running. If we are correct, it looks like this:

```
xcodebuild clean build CODE_SIGN_IDENTITY= CODE_SIGNING_REQUIRED=NO -target=Venue-companion
```

Next, we saw this guard statement and figured out that if `parseCompilerArguments` is `nil`, then report the error described originally by Jazzy:

```
guard let arguments = parseCompilerArguments(xcodeBuildOutput, language: .Swift, moduleName: name ?? moduleNameFromArguments(xcodeBuildArguments)) else { ... }
```

So then we searched the SourceKitten project for `parseCompilerArguments` and found [this file](https://github.com/jpsim/SourceKitten/blob/3dd448e5a70dd0a272b3541abc3355e046e9f59f/Source/SourceKittenFramework/Xcode.swift). We noticed that you are doing regex matching on the `xcodebuild` output for the module name:

```
...
if language == .ObjC {
   pattern = "/usr/bin/clang.*"
} else if let moduleName = moduleName {
   pattern = "/usr/bin/swiftc.*-module-name \(moduleName) .*"
} else {
   pattern = "/usr/bin/swiftc.*"
}
...
```

We ran the `xcodebuild` command described earlier and searched the output for something like the pattern `/usr/bin/swiftc.*-module-name \(moduleName) .*`. What we found was two instances of something that looks like this:

```
...
/Applications/Xcode7.1.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc -incremental -module-name Venue_companion
...
```

What's interesting here is that `Venue-companion`, after being pushed through `xcodebuild`, now has an underscore instead of a hyphen in it. The module name (presumably with a hyphen in it) is taken as an optional argument to `parseCompilerArguments`. When the pattern `/usr/bin/swiftc.*-module-name Venue-companion .*` is compared with the `xcodebuild` output later in `parseCompilerArguments`, that pattern is never found because the hyphen was replaced with an underscore, `nil` is returned, and Jazzy fails.

**Comment by jpsim** [comment]
Thanks for thorough report! We should definitely fix this.

**Comment by OkySabeni** [comment]
Is there a potential manual fix for this? My workspace and scheme has hyphens in it.

**Comment by trishayy** [comment]
I believe to fix this we had to rename our app... either without hyphens or changing the hyphens to underscores.

**Comment by OkySabeni** [comment]
Thanks @trishayy!

**Comment by JoshHrach** [comment]
@jpsim Was a fix for this ever implemented? I'm encountering this issue with Jazzy 0.7.2.

**Comment by istx25** [comment]
Let's organise who's going to implement this so we can get it in a new release. @JoshHrach, would you be interested in completing this?

**Comment by JoshHrach** [comment]
@istx25 I think what @trishayy mentioned above may be a place to start. I tried forking Jazzy to my system but am having issues setting up an environment.

What I also found was information from [this link](https://developer.apple.com/library/content/documentation/Swift/Conceptual/BuildingCocoaApps/MixandMatch.html) under **Naming Your Product Module**: 

> The name of the Xcode-generated header for Swift code, and the name of the Objective-C bridging header that Xcode creates for you, are generated from your product module name. By default, your product module name is the same as your product name. However, if your product name has any nonalphanumeric characters, such as a period (.), they are replaced with an underscore (_) in your product module name. If the name begins with a number, the first number is replaced with an underscore.

The project I'm working with here is an app for a brand with a hyphen. Going into Project > Build Settings > Packaging > Product Module Name, I was able to remove the hyphen and start using Jazzy again.

While that provides a workaround, it doesn't fix the initial issue with how Jazzy (or in this case, SourceKitten) is handling the Module Name.  As for an actual fix, until I can get it working in a local environment, I don't think I would be able to provide one. Ultimately, I think the line and the file that @trishayy pointed out is a place to start. `moduleName` needs to handle the situation in the above-mentioned link.

**Comment by johnfairh** [comment]
This is a bug in sourcekitten's inference of module-name from target-name, I guess should implement the algorithm quoted above.  Workaround is to explicitly pass the correct module name to jazzy.

**Comment by johnfairh** [comment]
This will eventually be fixed by jpsim/sourcekitten#578, thanks @chriszielinski!

**Comment by mobimation** [comment]
Does this cover the case where an iOS Swift app project is created in Xcode with a name containing "foreign alphabet characters ?  As an example here I here use the name "Amatör", a swedish representation of the english word "Amateur".  Then when building/running the app execution fails with "Compiling failed. No such module 'Amato_r'.'  The app project name letter "ö" becomes represented as "o_" in Xcode.
How do I work around this ?

**Comment by johnfairh** [comment]
I'm not sure if you're asking about Xcode or jazzy.  For jazzy purposes, looks like you'll need to write the module name explicitly: `jazzy --module Amatör`.

In Xcode11.3 by default, `PRODUCT_MODULE_NAME` is "Amat_r" and the Swift module is definitely called "Amatör".  This looks like an Xcode bug, the help for `PRODUCT_MODULE_NAME` hasn't changed.
